include/mesh/simplex_refiner.h

```diff
@@ -0,0 +1,198 @@
+   */
+  Real & desired_volume() {return _desired_volume;}
+
+protected:
+
+  /**
+   * Finds elements which exceed the requested metric and refines them
+   * via inserting new midedge nodes and bisecting as necessary.
+   *
+   * \returns the number of refined elements
+   */
+  std::size_t refine_via_edges();
+
+  /**
+   * Checks if an element exceeds the requested metric
+   */
+  bool should_refine_elem(Elem & elem);
+
+  /**
+   * Checks if an element exceeds the requested metric or if it
+   * has an edge which was split by a neighboring metric and refines
+   * it (bisecting it, removing it from the mesh to be replaced by the
+   * two subelements, and recursing into those) if necessary.
+   *
+   * The \p coarse_id specifies which coarse element this one is (or
+   * which this one was originally refined from), to make global
+   * communication easier after local refinement is done.
+   *
+   * \returns the number of refinements done; this may be greater than
+   * 1 if subelements were themselves refined.
+   */
+
+  std::size_t refine_via_edges(Elem & elem,
+                               dof_id_type coarse_id);
+
+private:
+
+  /**
+   * Reference to the mesh which is to be refined.
+   */
+  UnstructuredMesh & _mesh;
+
+  /**
+   * The desired volume for the elements in the resulting mesh.
+   */
+  Real _desired_volume;
+
+  /**
+   * Location-dependent volume requirements
+   */
+  std::unique_ptr<FunctionBase<Real>> _desired_volume_func;
+
+  /**
+   * Keep track of new nodes on edges.  A new node x in between node y
+   * and node z (with y<z) will be reflected by new_nodes[(y,z)]=x
+   */
+  std::map<std::pair<Node *, Node *>, Node *> new_nodes;
+
+  /**
+   * Keep track of elements to add so we don't invalidate iterators
+   * during an iteration over elements.
+   */
+  std::vector<std::unique_ptr<Elem>> new_elements;
+
+  /**
+   * Keep track of coarse remote edges for which we should query about
+   * additional point splits
+   */
+  std::unordered_map
+    <processor_id_type,
+     std::vector<std::pair<dof_id_type, dof_id_type>>>
+    edge_queries;
+
+  // All the "refine first and second to get third" that was
+  // done to this edge, including the processor to assign the new node
+  // to.
+  typedef
+    std::vector<std::tuple<dof_id_type, dof_id_type,
+                           dof_id_type, processor_id_type>>
+    refinement_datum;
+
+  /**
+   * Helper for responding to edge queries
+   */
+  void fill_refinement_datum(std::pair<Node *, Node *> vertices,
+                             refinement_datum & vec);
+
+  /**
+   * Keep track of elements added within each original element, so we
+   * can answer queries about them from other processors.  If the
+   * element originally with id i is split, its replacements e1 and e2
+   * will be in added_elements[i][ei.vertex_average()]
+   */
+  std::unordered_map<dof_id_type, std::unordered_map<Point, Elem *>> added_elements;
```

Comment:
Huh, I didn't realize we had a `std::hash<Point>` specialization. Is this how you want to handle it whenever we need to make an `unordered_map` with a non-trivial Key type? In #3975, I used the "class with operator()" approach, and this approach is also used in e.g. the `ElemHashUtils` class. Basically I'm just curious what is considered the best practice for user-defined hash functions... should they all just be put in namespace `std`?

Reply:
In #3975 you did the only legal thing, I think - our key is a `vector<dof_id_type>`, and `vector` isn't a user class, and `dof_id_type` is a user typedef but not a user class, so it's not legal to put our own specialization around it into namespace `std`.  With `Point`, because we created an actual class, doing our own specializations for it is specifically allowed ... though I'm not sure if it's considered best practice.  IMHO the main benefit of it is only that it makes generic algorithms easier if you don't have to figure out how to use a custom hash function template argument for some cases but not others.